File: table_results.py
import csv
import os
from collections import defaultdict
import textwrap

def generate_results_table(input_dir, output_file):
    """
    Lee archivos de resultados de experimentos y genera una tabla consolidada.
    """
    # 1. Definir los datasets y algoritmos
    datasets = ['Gov2', 'ClueWeb09', 'CC-News']
    algorithms = ['wBtrie', 'x2WBtrie', 'x2WRBtrie', 'x2WTRBtrie', 'x3WRBtrie', 'x3WTRBtrie']
    
    # Estructura para almacenar los datos
    results = defaultdict(lambda: defaultdict(dict))

    # 2. Leer los datos de cada fichero
    print("Leyendo archivos de resultados...")
    for dataset in datasets:
        for algo in algorithms:
            file_name = f"{algo}_{dataset}.csv"
            file_path = os.path.join(input_dir, file_name)

            if not os.path.exists(file_path):
                # print(f"Advertencia: No se encontró el archivo {file_path}. Se omite.")
                continue

            try:
                with open(file_path, 'r') as f:
                    reader = csv.DictReader(f)
                    
                    data_rows = list(reader)
                    
                    # Asumiendo 6 filas por archivo: 3 para v5 y 3 para v
                    if len(data_rows) >= 6:
                        # Extraer los datos de las columnas de interés
                        v5_data = [(float(row['avg_size_bits_per_element']), float(row['avg_time'])) for row in data_rows[:3]]
                        v_data = [(float(row['avg_size_bits_per_element']), float(row['avg_time'])) for row in data_rows[3:6]]
                        
                        results[algo][dataset]['v5'] = v5_data
                        results[algo][dataset]['v'] = v_data
            except Exception as e:
                print(f"Error al procesar el archivo {file_path}: {e}")

    # 3. Generar la tabla en un string
    print("Generando la tabla...")
    output_lines = []

    # Construir el encabezado de la tabla
    header = f"| Data Structure | {' &nbsp; '*10} | {' &nbsp; '*10} | {' &nbsp; '*10} |\n"
    header += f"|                | {'Gov2':^21} | {'ClueWeb09':^21} | {'CC-News':^21} |\n"
    header += f"|                | {'Space':>9} {'Time':>10} | {'Space':>9} {'Time':>10} | {'Space':>9} {'Time':>10} |"
    output_lines.append(header)
    output_lines.append("|:---|:---:|:---:|:---:|")

    # Llenar la tabla con los datos
    for algo in algorithms:
        if algo not in results:
            continue
        
        # Datos para la variante v5
        v5_data = [results[algo][ds].get('v5', [('N/A', 'N/A')] * 3) for ds in datasets]
        for i in range(3):
            line = f"| {algo} (v5) |"
            for j, dataset in enumerate(datasets):
                space, time = v5_data[j][i]
                line += f" {space:8.2f} {time:8.2f} |" if isinstance(space, float) else f" {'N/A':>8} {'N/A':>8} |"
            output_lines.append(line)

        # Datos para la variante v
        v_data = [results[algo][ds].get('v', [('N/A', 'N/A')] * 3) for ds in datasets]
        for i in range(3):
            line = f"| {algo} (v) |"
            for j, dataset in enumerate(datasets):
                space, time = v_data[j][i]
                line += f" {space:8.2f} {time:8.2f} |" if isinstance(space, float) else f" {'N/A':>8} {'N/A':>8} |"
            output_lines.append(line)
        output_lines.append("|---|---|---|---|")

    # 4. Escribir la tabla en el archivo de salida
    with open(output_file, 'w') as f:
        f.write("\n".join(output_lines))

    print(f"\n¡Tabla generada con éxito en '{output_file}'!")

# --- Uso del script ---
if __name__ == "__main__":
    # Asegúrate de que esta ruta apunte a la carpeta donde están tus 15 archivos CSV
    # Por ejemplo, si los archivos están en una carpeta llamada 'outputs' en el mismo directorio que el script
    input_directory = './outputs'
    output_file = 'tabla_resultados.md'
    
    generate_results_table(input_directory, output_file)

File: test/table_results.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <iomanip>

// Estructura para almacenar los datos de cada fila
struct ResultRow {
    double space;
    double time;
};

// Estructura para almacenar todos los resultados por variante de rank (v5 y v)
struct AlgoritmoResult {
    std::vector<ResultRow> v5_data;
    std::vector<ResultRow> v_data;
};

// Función para leer un archivo CSV y extraer los datos
AlgoritmoResult read_csv_file(const std::string& filepath) {
    AlgoritmoResult result;
    std::ifstream file(filepath);
    if (!file.is_open()) {
        std::cerr << "Advertencia: No se pudo abrir el archivo " << filepath << std::endl;
        return result;
    }

    std::string line;
    std::getline(file, line); // Leer y descartar la línea de encabezado

    std::vector<ResultRow> all_rows;
    while (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string cell;
        ResultRow row;
        
        // Asumiendo que las columnas de interés son la 3 y la 5
        for (int i = 0; std::getline(ss, cell, ','); ++i) {
            if (i == 2) {
                row.space = std::stod(cell);
            }
            if (i == 4) {
                row.time = std::stod(cell);
            }
        }
        all_rows.push_back(row);
    }
    
    // Asumiendo que las primeras 3 filas son para v5 y las siguientes 3 para v
    if (all_rows.size() >= 6) {
        result.v5_data = std::vector<ResultRow>(all_rows.begin(), all_rows.begin() + 3);
        result.v_data = std::vector<ResultRow>(all_rows.begin() + 3, all_rows.begin() + 6);
    }

    return result;
}

int main() {
    // Definir los datasets y algoritmos
    std::vector<std::string> datasets = {"Gov2", "ClueWeb09", "CC-News"};
    std::vector<std::string> algorithms = {"wBtrie", "x2WBtrie", "x2WRBtrie", "x2WTRBtrie", "x3WRBtrie", "x3WTRBtrie"};
    std::string input_dir = "outputs/"; // La carpeta donde están los archivos CSV
    std::string output_file = "tabla_resultados.md";
    
    // Estructura para almacenar todos los datos leídos
    std::map<std::string, std::map<std::string, AlgoritmoResult>> results;

    std::cout << "Leyendo archivos de resultados..." << std::endl;
    for (const auto& algo : algorithms) {
        for (const auto& dataset : datasets) {
            std::string filename = algo + "_" + dataset + ".csv";
            std::string filepath = input_dir + filename;
            AlgoritmoResult data = read_csv_file(filepath);
            
            if (!data.v5_data.empty() || !data.v_data.empty()) {
                results[algo][dataset] = data;
            }
        }
    }

    // Generar la tabla en un string
    std::cout << "Generando la tabla..." << std::endl;
    std::stringstream table_ss;

    // Encabezado
    table_ss << "| Data Structure |"
             << std::setw(23) << std::left << "Gov2" << "|"
             << std::setw(23) << std::left << "ClueWeb09" << "|"
             << std::setw(23) << std::left << "CC-News" << "|" << std::endl;
    
    table_ss << "|                | "
             << "Space      Time | "
             << "Space      Time | "
             << "Space      Time |" << std::endl;
    
    table_ss << "|:---|:---:|:---:|:---:|" << std::endl;

    // Llenar la tabla con los datos
    for (const auto& algo : algorithms) {
        if (results.find(algo) == results.end()) {
            continue;
        }
        
        // Variante v5
        for (size_t i = 0; i < 3; ++i) {
            table_ss << "| " << algo << " (v5) |";
            for (const auto& dataset : datasets) {
                if (results[algo].count(dataset) && results[algo][dataset].v5_data.size() > i) {
                    table_ss << std::fixed << std::setprecision(2) << std::setw(8) << results[algo][dataset].v5_data[i].space << " "
                             << std::setw(8) << results[algo][dataset].v5_data[i].time << " |";
                } else {
                    table_ss << std::setw(9) << "N/A" << std::setw(8) << "N/A" << " |";
                }
            }
            table_ss << std::endl;
        }

        // Variante v
        for (size_t i = 0; i < 3; ++i) {
            table_ss << "| " << algo << " (v) |";
            for (const auto& dataset : datasets) {
                if (results[algo].count(dataset) && results[algo][dataset].v_data.size() > i) {
                    table_ss << std::fixed << std::setprecision(2) << std::setw(8) << results[algo][dataset].v_data[i].space << " "
                             << std::setw(8) << results[algo][dataset].v_data[i].time << " |";
                } else {
                    table_ss << std::setw(9) << "N/A" << std::setw(8) << "N/A" << " |";
                }
            }
            table_ss << std::endl;
        }
        table_ss << "|---|---|---|---|" << std::endl;
    }

    // Escribir la tabla al archivo de salida
    std::ofstream out_file(output_file);
    if (out_file.is_open()) {
        out_file << table_ss.str();
        out_file.close();
        std::cout << "\n¡Tabla generada con éxito en '" << output_file << "'!" << std::endl;
    } else {
        std::cerr << "Error: No se pudo crear el archivo de salida '" << output_file << "'" << std::endl;
    }

    return 0;
}